libcudacxx/include/cuda/try_cancel

```diff
@@ -0,0 +1,324 @@
+ * This license may only be modified in a writing signed by an authorized representative of each party.
+ *
+ * (v. August 20, 2021)
+ */
+
+#ifndef _CUDA_TRY_CANCEL
+#define _CUDA_TRY_CANCEL
+
+#include <cuda/std/detail/__config>
+
+#if defined(_CCCL_IMPLICIT_SYSTEM_HEADER_GCC)
+#  pragma GCC system_header
+#elif defined(_CCCL_IMPLICIT_SYSTEM_HEADER_CLANG)
+#  pragma clang system_header
+#elif defined(_CCCL_IMPLICIT_SYSTEM_HEADER_MSVC)
+#  pragma system_header
+#endif // no system header
+
+_LIBCUDACXX_BEGIN_NAMESPACE_CUDA
+
+namespace experimental {
+
+namespace __detail {
+
+struct __empty_t {
+  __device__ void operator()(dim3);
+};
+
+template <int __I>
+__device__ int __cluster_get_dim(__int128 __result) {
+    int __r;
+    if constexpr(__I == 0) {
+        asm volatile("clusterlaunchcontrol.query_cancel.get_first_ctaid::x.b32.b128 %0, %1;" : "=r"(__r) : "q"(__result) : "memory");
+    } else if constexpr(__I == 1) {
+        asm volatile("clusterlaunchcontrol.query_cancel.get_first_ctaid::y.b32.b128 %0, %1;" : "=r"(__r) : "q"(__result) : "memory");
+    } else if constexpr(__I == 2) {
+        asm volatile("clusterlaunchcontrol.query_cancel.get_first_ctaid::z.b32.b128 %0, %1;" : "=r"(__r) : "q"(__result) : "memory");
+    } else {
+      __r = -1;
+      #ifndef NDEBUG
+      __trap();
+      #endif
+    }
+    return __r;
+}
+
+} // namespace __detail
+
+/// This API for implementing work-stealing, repeatedly attempts to cancel the launch of a thread block
+/// from the current grid. On success, it invokes the unary function `__uf` before trying again.
+/// On failure, it returns.
+///
+/// This API does not provide any memory synchronization.
+/// This API does not guarantee that any thread will invoke `__uf` with the next block index until all
+/// invocatons of `__uf` for the prior block index have returned.
+///
+/// Preconditions:
+/// - All thread block threads shall call this API exactly once.
+/// - Exactly one thread block thread shall call this API with `__is_leader` equals `true`.  
+template <int __ThreadBlockDim = 3, typename __UnaryFunction = __detail::__empty_t>
+    requires std::is_invocable_r_v<void, __UnaryFunction, dim3>
+__device__ void try_cancel_blocks(bool __is_leader, __UnaryFunction __uf) {
+  #if (__CUDA_ARCH__ < 1000)
+    __trap();
+  #else   // NV_HAS_FEATURE_SM100
+    __shared__ uint64_t __barrier; // TODO: use 2 barriers and 2 results avoid last sync threads
+    __shared__ __int128 __result;
+    bool __phase = false;
+    dim3 __block_idx;
+
+    if constexpr (__ThreadBlockDim == 3) {
+      __block_idx = dim3(blockIdx.x, blockIdx.y, blockIdx.z);
+    } else if constexpr (__ThreadBlockDim == 2) {
+      __block_idx = dim3(blockIdx.x, blockIdx.y, 1);
+    } else if constexpr (__ThreadBlockDim == 1) {
+      __block_idx = dim3(blockIdx.x, 1, 1);      
+    } else {
+      #ifndef NDEBUG
+        __trap();
+      #endif
```

Comment:
ditto should rather be `_CUDA_VSTD::terminate()`

Reply:
These are now `static_assert`.